### PR TITLE
Copy container logs as single tar archive

### PR DESCRIPTION
**Pull request by wendigo**

### CODE REVIEW

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java [hunk 1]
```diff
@@ -18,6 +18,7 @@
 import com.github.dockerjava.api.model.HealthCheck;
```
Comment (kokosing) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java:18]: What is the motivation?

**Comment by findepi** [comment]
Did we reach the point where Launcher is so complex it should come with battery of tests?

**Comment by wendigo** [comment]
> Did we reach the point where Launcher is so complex it should come with battery of tests?
> 
> 

I was thinking about testing launcher for some time but complexity of it arises from shortcomings of testcontainers abd dockerjava

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java [hunk 2]
```diff
@@ -201,37 +206,67 @@ public void copyLogsToHostPath(Path hostPath)
         Path hostLogPath = Paths.get(hostPath.toString(), logicalName);
         ensurePathExists(hostLogPath);
 
+        ImmutableList.Builder<String> files = ImmutableList
+                .builder();
```
Comment (kokosing) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java:218]: no point of splitting the line here

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java [hunk 3]
```diff
@@ -182,11 +186,23 @@ protected void containerIsStopped(InspectContainerResponse containerInfo)
...
+            executor.runAsync(copy).whenComplete((ignore, throwable) -> {
+                if (throwable == null) {
+                    log.info("Copied files into %s %s in %.1f s", this, containerPath, stopwatch.elapsed(MILLISECONDS) / 1000.);
+                }
+                else {
+                    log.warn("Could not copy files into %s %s: %s", this, containerPath, getStackTraceAsString((Throwable) throwable));
+                }
+            }).get();
+        }
+        catch (Throwable e) {
```
Comment (kokosing) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java:195]: Throwable?

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java [hunk 4]
```diff
@@ -23,10 +23,12 @@
 import com.google.common.base.Stopwatch;
```
Comment (kokosing) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java:23]: > Make more operations async

What is the benefit of this?

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java [hunk 5]
```diff
@@ -370,7 +386,7 @@ public void tryStop()
         }
 
         try {
-            executor.run(this::stop);
+            executor.runAsync(this::stop).get();
```
Comment (kokosing) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java:379]: What is the difference between `run(...)` and  `runAsync(...).get()`?

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/Environment.java [hunk 6]
```diff
@@ -80,6 +83,9 @@
     private final Map<String, DockerContainer> containers;
     private final Optional<EnvironmentListener> listener;
 
+    private static final ExecutorService executorService = Executors
+            .newCachedThreadPool(new ThreadFactoryBuilder().setNameFormat("environment-%d").build());
```
Comment (kokosing) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/Environment.java:87]: io.airlift.concurrent.Threads#daemonThreadsNamed?

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java (hunk 4 above)
Comment (wendigo) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java:23]: Previously some of the operations were not interruptible + they were executed in thread pool not controlled by launcher (okhttp from testcontainers)

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/Environment.java (hunk 6 above)
Comment (wendigo) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/Environment.java:87]: Didn't know that :) What's the difference?

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java (hunk 5 above)
Comment (wendigo) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java:379]: runAsync executed CheckedRunnable on scheduler

Comment (losipiuk) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java:379]: ?

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java [hunk 7]
```diff
@@ -302,10 +302,8 @@ private void copyFileFromContainer(String filename, Path targetPath)
             return Optional.empty();
         }
 
-        try (DockerClient client = DockerClientFactory.lazyClient()) {
-            InvocationBuilder.AsyncResultCallback<Statistics> callback = new InvocationBuilder.AsyncResultCallback<>();
+        try (DockerClient client = DockerClientFactory.lazyClient(); AsyncResultCallback<Statistics> callback = new AsyncResultCallback<>()) {
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java:305]: Seems like the library is broken. I would expect the `DockerClient.close` to close the callback passed to `client.statsCmd(getContainerId()).exec(callback);`.
Did you maybe report a bug?

**Review by losipiuk — APPROVED** [review]
LGTM

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java (hunk 7 above)
Comment (wendigo) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java]: No. I didn't. And I wouldn't expect that to be fixed like you've described.

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/StatisticsFetcher.java [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+        return stats;
+    }
+
+    public static class Stats
+    {
+        private long systemCpuUsage = -1;
+        private long totalCpuUsage = -1;
+        private double cpuUsagePerc;
+        private double memoryUsagePerc;
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/StatisticsFetcher.java:111]: nit: do not use abbreviation

**Review by losipiuk — APPROVED** [review]

**Pull request merged**